### PR TITLE
fix: Resolve Content-Length header conflict with Transfer-Encoding: chunked

### PR DESCRIPTION
**Pull request by klkucaj**
This PR resolves an issue where `Content-Length` was incorrectly included alongside `Transfer-Encoding: chunked`.

**Fixes Implemented:**
- Removed `Content-Length` when `Transfer-Encoding: chunked` is applied.
- Ensured `Content-Length` is only included for other valid `Transfer-Encoding` values (e.g., gzip, deflate).
- Added logic to reorder transfer encodings, ensuring `chunked` is always the last encoding.

**Impact:**
This fix ensures compliance with HTTP/1.1, preventing ambiguity and avoiding potential protocol errors in HTTP clients and servers.

**Closes:** #20

### CODE REVIEW

File: test/relic_server_serve_test.dart [hunk 1]
```diff
@@ -499,7 +499,7 @@ void main() {
         response.headers,
         containsPair(HttpHeaders.transferEncodingHeader, 'chunked'),
       );
-      expect(response.body, equals('hi'));
+      expect(response.body, equals('2\r\nhi\r\n0\r\n\r\n'));
```
Comment (SandPod) [inline test/relic_server_serve_test.dart:502]: Suggestion: Could you clarify the description on this test so that it is clear what and why we do this validation?

File: test/message/body_test.dart [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+          headers: Headers.response(
+            transferEncoding: TransferEncodingHeader(
+              encodings: [TransferEncoding.chunked],
+            ),
+          ),
+        ),
+      );
+
+      var response = await _get();
+      expect(response.headers['transfer-encoding'], contains('chunked'));
```
Comment (SandPod) [inline test/message/body_test.dart:128]: Suggestion: Use `TransferEncoding.chunked` in the validation for the tests.

Comment (SandPod) [reply test/message/body_test.dart:128]: Use this for all tests when referring to any header string.

**Review by SandPod — CHANGES_REQUESTED** [review]
I know a lot of digging went into this one so it will be nice to get this resolved.

I had a couple of comments and suggestion. Maybe we could have a call if anything is unclear.

File: test/headers/headers_test_utils.dart [hunk 3]
```diff
@@ -17,13 +17,16 @@ class BadRequestException implements Exception {
...
+  bool echoHeaders = true,
 }) async {
   Headers? parsedHeaders;
 
   server.mountAndStart(
     (Request request) {
       parsedHeaders = request.headers;
-      return Response.ok();
+      return Response.ok(
+        headers: echoHeaders ? parsedHeaders : null,
```
[the remark this answers is not in the document]
Comment (klkucaj) [reply test/headers/headers_test_utils.dart:28]: No, there's no difference. Whether you explicitly set `headers` to `null` or just call `Response.ok()` without modifying it, the outcome is the same. The constructor handles it like this:  

```dart
...
headers: headers ?? Headers.response(),
...
```

This means if `headers` is `null`, it will default to `Headers.response()`.

File: lib/src/headers/typed/headers/transfer_encoding_header.dart [hunk 4]
```diff
@@ -42,6 +43,35 @@ class TransferEncodingHeader implements TypedHeader {
...
+  ///
+  /// This function reorders the encodings to comply with the standard and
+  /// ensures compatibility with HTTP clients and intermediaries.
+  static List<TransferEncoding> _reorderEncodings(
+    List<TransferEncoding> encodings,
+  ) {
+    final TransferEncoding? chunked = encodings.firstWhereOrNull(
+      (e) => e.name == TransferEncoding.chunked.name,
+    );
+    if (chunked == null) return encodings;
```
[the remark this answers is not in the document]
Comment (klkucaj) [reply lib/src/headers/typed/headers/transfer_encoding_header.dart:68]: The reason I used `firstWhereOrNull` instead of `.contains` is that `.contains` only works directly with primitive types or objects that have properly overridden the `==` operator and `hashCode`. 
In this case, since `TransferEncoding` is a custom class and doesn’t inherently support those overrides, `.contains` wouldn’t work out of the box without additional implementation.

By using `firstWhereOrNull`, I can avoid the need to override `==` and `hashCode`, and directly match based on a specific property (`name` in this case), which keeps the logic simple and avoids unnecessary changes to the `TransferEncoding` class.

File: lib/src/headers/headers.dart [hunk 5]
```diff
@@ -1237,10 +1235,7 @@ final class _HeadersImpl extends Headers {
 
   /// Apply headers to the response
   @override
-  void applyHeaders(
-    io.HttpResponse response,
-    Body body,
-  ) {
+  void applyHeaders(io.HttpResponse response) {
```
[the remark this answers is not in the document]
Comment (klkucaj) [reply lib/src/headers/headers.dart:1238]: Yes, you are right!

File: lib/src/body/types/mime_type.dart [hunk 6]
```diff
@@ -33,6 +33,12 @@ class MimeType {
   /// RTF mime type.
   static const rtf = MimeType('application', 'rtf');
 
+  /// Multipart mime type.
+  static const multipart = MimeType('multipart', 'form-data');
```
Comment (SandPod) [inline lib/src/body/types/mime_type.dart:37]: Suggestion: Rename to `formData`.

Following the naming used above i would expect this to be named form data? What does dart use?

File: lib/src/body/types/mime_type.dart [hunk 7]
```diff
@@ -33,6 +33,12 @@ class MimeType {
...
+
+  /// Multipart mime type.
+  static const multipartByteranges = MimeType('multipart', 'byteranges');
```
Comment (SandPod) [inline lib/src/body/types/mime_type.dart:40]: Suggestion: Following how `plainText` is named, this should probably be named `byterangesMultipart` right? What does dart call these?

File: lib/src/body/types/mime_type.dart [hunk 8]
```diff
@@ -74,3 +80,13 @@ extension ContentTypeExtension on ContentType {
   /// We are calling this method 'toMimeType' to avoid conflict with the 'mimeType' property.
   MimeType get toMimeType => MimeType(primaryType, subType);
 }
+
+/// Extension to check if a [MimeType] is not multipart/byteranges.
+extension MimeTypeExtensions on MimeType {
+  /// Checks if the mime type is not multipart/byteranges.
+  bool get isMultipartByteranges {
```
Comment (SandPod) [inline lib/src/body/types/mime_type.dart:87]: Suggestion: If this is only used in a single place, do the check there directly insteaaed of adding an extention.

File: lib/src/body/body.dart [hunk 9]
```diff
@@ -3,7 +3,7 @@ import 'dart:convert';
 import 'dart:io';
 import 'dart:typed_data';
 
-import 'package:relic/src/body/types/body_type.dart';
+import 'package:relic/relic.dart';
```
Comment (SandPod) [inline lib/src/body/body.dart:6]: Question: This file now only contains a change to this import, is this required?

File: lib/src/extensions/http_response_extension.dart [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+      httpHeaders.set(key, value);
+    }
+
+    // Set the Content-Type header based on the MIME type of the body.
+    httpHeaders.contentType = body.getContentType();
+
+    // If the content length is known, set it and remove the Transfer-Encoding header.
+    if (body.contentLength != null) {
+      httpHeaders
+        ..contentLength = body.contentLength!
```
Comment (SandPod) [inline lib/src/extensions/http_response_extension.dart:27]: Suggestion: Assign the content length before nullcheck so that the dangerous derreff can be avoided ( `!`)

File: lib/src/extensions/http_response_extension.dart [hunk 11]
```diff
@@ -0,0 +1,68 @@
+import 'package:relic/relic.dart';
+import 'dart:io' as io;
+
+import 'package:relic/src/body/types/mime_type.dart';
+
+/// Extension for [io.HttpResponse] to apply headers and body.
+extension HttpResponseExtension on io.HttpResponse {
+  /// Apply headers and body to the response.
+  void applyHeaders(Headers headers, Body body) {
+    var httpHeaders = this.headers;
```
Comment (SandPod) [inline lib/src/extensions/http_response_extension.dart:10]: Suggestion: Rename to `responseHeaders` to make it clear what this is.

File: lib/src/extensions/http_response_extension.dart [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    httpHeaders.clear();
+
+    // Apply headers
+    var mappedHeaders = headers.toMap();
+    for (var entry in mappedHeaders.entries) {
+      var key = entry.key;
+      var value = entry.value;
```
Comment (SandPod) [inline lib/src/extensions/http_response_extension.dart:17]: Suggestion(NIT): Skip assigning this variable and just pass it directly to the set method.

`httpHeaders.set(entry.key, entry.value);`

File: lib/src/extensions/http_response_extension.dart [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+      // Set Transfer-Encoding header and remove Content-Length as it is not needed for chunked encoding.
+      httpHeaders
+        ..set(
+          Headers.transferEncodingHeader,
+          encodings.map((e) => e.name).toList(),
+        )
+        ..removeAll(Headers.contentLengthHeader);
+    } else {
+      // Set Content-Length to 0 if chunked encoding is not enabled.
+      httpHeaders.contentLength = 0;
```
Comment (SandPod) [inline lib/src/extensions/http_response_extension.dart:65]: Question: If we don't use chunked encoding, shouldn't this be the correct content length?

File: test/relic_server_serve_test.dart [hunk 14]
```diff
@@ -508,6 +508,7 @@ void main() {
           return Response.ok(
             body: Body.fromDataStream(
               Stream.value(Uint8List.fromList([1, 2, 3, 4])),
+              contentLength: 4,
```
[the remark this answers is not in the document]
Comment (SandPod) [reply test/relic_server_serve_test.dart:525]: No reply and no fix to this comment, Please look at it again.

**Review by SandPod — COMMENTED** [review]
Couple of more comment and suggestion.

File: lib/src/extensions/http_response_extension.dart (hunk 13 above)
Comment (klkucaj) [reply lib/src/extensions/http_response_extension.dart:65]: Your right.

File: lib/src/body/body.dart (hunk 9 above)
Comment (klkucaj) [reply lib/src/body/body.dart:6]: I am going to revert this to main because there no need to do any changes here, the reason why is only this because this is how it is on main https://github.com/klkucaj/relic/blob/c4a9eedd99636336dbf2ce7338b1ccb574d4205b/lib/src/body/body.dart#L6-L7

File: test/relic_server_serve_test.dart (hunk 14 above)
[the remark this answers is not in the document]
Comment (klkucaj) [reply test/relic_server_serve_test.dart:525]: Yes u are right, sorry, i improved other tests and i missed this.

File: lib/src/body/types/mime_type.dart (hunk 7 above)
Comment (klkucaj) [reply lib/src/body/types/mime_type.dart:40]: @SandPod check the new version of mime type and let me know if its okay.

File: test/message/body_test.dart [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+      var response = await _get();
+      expect(
+        response.headers['transfer-encoding'],
+        contains(TransferEncoding.chunked.name),
+      );
+      expect(response.body, equals('5\r\nRelic\r\n0\r\n\r\n'));
+    });
+
+    test(
+        'with a valid content length when applying headers '
```
Comment (SandPod) [inline test/message/body_test.dart:143]: Suggestion: Add a line break for each section in the `Given, When, Then` pattern.

This should be applied to all tests in this file.

For this test that would result in:

```dart
'with a valid content length '
'when applying headers to the response '
'then Content-Length is used instead of chunked encoding'
```

File: test/message/body_test.dart [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+        Completer<Object?> completer = Completer();
+
+        unawaited(runZonedGuarded(
+          () async {
+            await _scheduleServer(
+              (_) => Response.ok(
+                body: Body.fromDataStream(
+                  Stream.fromIterable([
+                    Uint8List.fromList([1, 2, 3, 4])
+                  ]),
```
Comment (SandPod) [inline test/message/body_test.dart:178]: Question: What happens if no data is returned from the stream?

File: lib/src/extensions/http_response_extension.dart [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+    var contentLength = body.contentLength;
+    if (contentLength != null) {
+      responseHeaders
+        ..contentLength = contentLength
+        ..removeAll(Headers.transferEncodingHeader);
+      return;
+    }
+
+    // Check if the content type is multipart/byteranges.
+    var bodyMimeType = body.contentType?.mimeType;
```
Comment (SandPod) [inline lib/src/extensions/http_response_extension.dart:30]: Question: What is the difference between `body.contentType` and `body.getContentType`?

File: lib/src/extensions/http_response_extension.dart [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+        // 304 is not modified
+        statusCode != 304 &&
+        // If the content type is not multipart/byteranges, chunked encoding is applied.
+        !isMultipartByteranges;
+
+    // Prepare transfer encodings.
+    var encodings = headers.transferEncoding?.encodings ?? [];
+    bool isChunked = headers.transferEncoding?.isChunked ?? false;
+
+    if (shouldEnableChunkedEncoding && !isChunked) {
```
Comment (SandPod) [inline lib/src/extensions/http_response_extension.dart:48]: Discussion: This still looks a bit funky to me. Let's discuss it.

File: lib/src/extensions/http_response_extension.dart (hunk 17 above)
Comment (klkucaj) [reply lib/src/extensions/http_response_extension.dart]: Renamed `body.contentType` to `body.bodyType`, this is our implementation of `contentType`.

File: test/message/body_test.dart (hunk 16 above)
Comment (klkucaj) [reply test/message/body_test.dart]: This is removed now, we dont need it anymore.

File: lib/src/extensions/http_response_extension.dart [hunk 19]
```diff
@@ -0,0 +1,83 @@
+import 'package:relic/relic.dart';
+import 'dart:io' as io;
+
+/// Extension for [io.HttpResponse] to apply headers and body.
+extension HttpResponseExtension on io.HttpResponse {
+  /// Apply headers and body to the response.
+  ///
+  /// Resolves conflicts between `Content-Length` and `Transfer-Encoding`
```
Comment (SandPod) [inline lib/src/extensions/http_response_extension.dart:8]: Suggestion: Clarify this comment. We no longer resolve conflicts but simply add the the chunked header if required by the body.

Comment (SandPod) [reply lib/src/extensions/http_response_extension.dart:8]: Combine this with the comment above as well.

File: lib/src/headers/typed/headers/transfer_encoding_header.dart [hunk 20]
```diff
@@ -33,6 +34,11 @@ class TransferEncodingHeader implements TypedHeader {
     return encodings.contains(TransferEncoding.chunked);
   }
 
+  /// Checks if the Transfer-Encoding contains `identity`.
+  bool get isIdentity {
```
Comment (SandPod) [inline lib/src/headers/typed/headers/transfer_encoding_header.dart:38]: Suggestion: Is `isChunked` and `isIdentity` only used in apply headers? If so, these could be extensions there instead of bloating this class.

File: lib/src/message/message.dart [hunk 21]
```diff
@@ -20,19 +20,19 @@ abstract class Message {
     this.context = const {},
   });
 
-  /// Returns the MIME type from the Content-Type header, if available.
-  MimeType? get mimeType => body.contentType?.mimeType;
+  /// Returns the MIME type from the Body-Type (Content-Type) header, if available.
```
Comment (SandPod) [inline lib/src/message/message.dart:23]: Suggestion: Should `header` be inside of the `)` as well?

File: lib/src/message/response.dart [hunk 22]
```diff
@@ -127,6 +125,22 @@ class Response extends Message {
           context: context,
         );
 
+  /// Constructs a 204 No Content response.
+  ///
+  /// This indicates that the request has succeeded but that the server has no
+  /// further information to send in the response body.
+  ///
+  /// {@macro relic_response_body_and_encoding_param}
```
Comment (SandPod) [inline lib/src/message/response.dart:133]: Question: What is the macro here?

File: test/relic_server_serve_test.dart [hunk 23]
```diff
@@ -448,43 +448,28 @@ void main() {
...
-            body: Body.fromDataStream(
-                Stream.value(Uint8List.fromList([1, 2, 3, 4]))),
-          );
-        });
-
-        var response = await _get();
-        expect(response.headers,
-            containsPair(HttpHeaders.transferEncodingHeader, 'chunked'));
-        expect(response.bodyBytes, equals([1, 2, 3, 4]));
+    test('is added when the transfer-encoding header is unset', () async {
```
Comment (SandPod) [inline test/relic_server_serve_test.dart:451]: Suggestion: The description here doesn't really explain the difference in the tests here. The important part is that we don't have a content length in order for the chunked to be added.

Comment (SandPod) [reply test/relic_server_serve_test.dart:451]: Same for the other tests in this group.

File: test/message/body_test.dart [hunk 24]
```diff
@@ -0,0 +1,193 @@
...
+import 'dart:io';
+import 'dart:typed_data';
+
+import 'package:http/http.dart' as http;
+import 'package:relic/relic.dart';
+import 'package:relic/src/method/request_method.dart';
+import 'package:relic/src/relic_server_serve.dart' as relic_server;
+import 'package:test/test.dart';
+
+void main() {
```
Comment (SandPod) [inline test/message/body_test.dart:11]: Question: Are these the same as the tests in `relic_server_serve_test`?

File: lib/src/message/response.dart (hunk 22 above)
Comment (klkucaj) [reply lib/src/message/response.dart:133]: A way to share comments that are only written in one place.

File: test/message/body_test.dart (hunk 24 above)
Comment (klkucaj) [reply test/message/body_test.dart]: Fixed!

File: test/relic_server_serve_test.dart (hunk 23 above)
Comment (klkucaj) [reply test/relic_server_serve_test.dart]: I updated some of the tests, moving a portion to `apply_headers_test.dart`. However, I didn't update the others as they were directly imported from Shelf. I'm unsure if updating these is necessary at this point.
What do u suggest ?

File: test/message/apply_headers_test.dart [hunk 25]
```diff
@@ -0,0 +1,296 @@
...
+
+  @override
+  HttpHeaders get headers => _headers;
+
+  @override
+  int get statusCode => _statusCode;
+}
+
+void main() {
+  group('Given a response', () {
```
Comment (SandPod) [inline test/message/apply_headers_test.dart:55]: Suggestion(NIT): For clarity: `Given a framework response object with some headers when applying headers to http response then headers are applied correctly`?

File: test/message/apply_headers_test.dart [hunk 26]
```diff
@@ -0,0 +1,296 @@
...
+  }) : _statusCode = statusCode;
+
+  @override
+  HttpHeaders get headers => _headers;
+
+  @override
+  int get statusCode => _statusCode;
+}
+
+void main() {
```
Comment (SandPod) [inline test/message/apply_headers_test.dart:54]: Suggestion: We are missing one test here now right, the test when we actually add the transfer chunked encoded? When content-lenght = null.

**Review by SandPod — COMMENTED** [review]
Last fix!

File: test/message/apply_headers_test.dart (hunk 25 above)
Comment (klkucaj) [reply test/message/apply_headers_test.dart]: Done

**Review by SandPod — APPROVED** [review]
LGTM! 👍
